packages/itwinui-react/src/core/Dialog/DialogMain.tsx

```diff
@@ -158,6 +161,33 @@ export const DialogMain = React.forwardRef((props, ref) => {
     }));
   }, []);
 
+  const onEnter = React.useCallback(() => {
+    previousFocusedElement.current = dialogRef.current?.ownerDocument
+      .activeElement as HTMLElement;
+    setFocus && dialogRef.current?.focus({ preventScroll: true });
+  }, [dialogRef, previousFocusedElement, setFocus]);
+
+  /** Brings back focus to the previously focused element when closed. */
+  const beforeClose = React.useCallback(() => {
+    if (
+      dialogRef.current?.contains(
+        dialogRef.current?.ownerDocument.activeElement,
+      )
+    ) {
+      previousFocusedElement.current?.focus();
+    }
+  }, [dialogRef, previousFocusedElement]);
+
+  const mountRef = React.useCallback(
+    (element: HTMLElement | null) => {
+      /** Focuses dialog when opened. */
+      if (element) {
+        onEnter();
+      }
+    },
+    [onEnter],
+  );
```

Comment:
nit 2: This can be made more readable by moving the comment and using an if-condition.

```suggestion
  /** Focuses dialog when opened. */
  const onEnter = React.useCallback(() => {
    previousFocusedElement.current = dialogRef.current?.ownerDocument
      .activeElement as HTMLElement;
    if (setFocus) {
      dialogRef.current?.focus({ preventScroll: true });
    }
  }, [dialogRef, previousFocusedElement, setFocus]);

  /** Brings back focus to the previously focused element when closed. */
  const beforeClose = React.useCallback(() => {
    if (
      dialogRef.current?.contains(
        dialogRef.current?.ownerDocument.activeElement,
      )
    ) {
      previousFocusedElement.current?.focus();
    }
  }, [dialogRef, previousFocusedElement]);

  const mountRef = React.useCallback(
    (element: HTMLElement | null) => {
      if (element) {
        onEnter();
      }
    },
    [onEnter],
  );
```